### PR TITLE
feat: support to enable quic and add basic test case

### PR DESCRIPTION
**Pull request by zll600**
### Description

Fixes https://github.com/apache/apisix/issues/6403

support enable http3 for apisix.

I will add a how-to guide in another PR.

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: ci/linux_install_static_curl.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+set -ex
+
+# use curl which supports http3 to test.
+install_static_curl
```
Comment (zll600) [inline ci/linux_install_static_curl.sh:22]: curl from apt does not support option --http3-only to test. So we need a newer version of curl

File: t/APISIX.pm [hunk 2]
```diff
@@ -777,7 +779,7 @@ _EOC_
             set \$upstream_connection         '';
 
             set \$upstream_scheme             'http';
-            set \$upstream_host               \$http_host;
+            set \$upstream_host               \$host;
```
Comment (zll600) [inline t/APISIX.pm:782]: The value of variable $http_host header will be empty string when using http/3
ref: https://github.com/zulip/zulip/pull/26348

File: apisix/cli/ngx_tpl.lua [hunk 3]
```diff
@@ -719,7 +722,7 @@ http {
             set $upstream_connection         '';
 
             set $upstream_scheme             'http';
-            set $upstream_host               $http_host;
+            set $upstream_host               $host;
```
Comment (zll600) [inline apisix/cli/ngx_tpl.lua:725]: The value of variable $http_host header will be empty string when using http/3
ref: zulip/zulip#26348

File: conf/config-default.yaml [hunk 4]
```diff
@@ -97,9 +97,11 @@ apisix:
     listen:                                       # APISIX listening port for HTTPS traffic.
       - port: 9443
         enable_http2: true
+        enable_quic: true
```
Comment (membphis) [inline conf/config-default.yaml:100]: I don't think we should use `true` as default
`false` is better

File: t/node/grpc-proxy.t [hunk 5]
```diff
@@ -219,7 +219,7 @@ grpcurl -import-path ./t/grpc_server_example/proto -proto helloworld.proto -plai
 --- grep_error_log eval
 qr/grpc header: "(:authority|host): [^"]+"/
 --- grep_error_log_out eval
-qr/grpc header: "(:authority|host): 127.0.0.1:1984"/
+qr/grpc header: "(:authority|host): 127.0.0.1"/
```
Comment (membphis) [inline t/node/grpc-proxy.t:222]: I don't think we need to update this line

File: t/node/http_host.t [hunk 6]
```diff
@@ -64,5 +64,5 @@ GET /uri
 Host: foo.com:1984
 --- response_body
 uri: /uri
-host: foo.com:1984
+host: foo.com
```
Comment (membphis) [inline t/node/http_host.t:67]: same issue, I do not know why we need to update this test case

File: t/quic/admin/basic.t [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        local code, body = t('/apisix/admin/routes/1',
+            ngx.HTTP_PUT,
+            [[{
+                "upstream": {
+                    "nodes": {
+                        "127.0.0.1:1980": 1
+                    },
+                    "type": "roundrobin"
+                },
+                "uris": ["/hello", "/world"]
```
Comment (membphis) [inline t/quic/admin/basic.t:86]: only `/hello` should be enough

File: t/node/http_host.t (hunk 6 above)
Comment (zll600) [reply t/node/http_host.t:67]: I change `$http_host` to `$host`, ref https://github.com/apache/apisix/pull/10989/files#r1510533390
`$host` has no `port`. So I will lost the port info here.

File: t/quic/admin/basic.t (hunk 7 above)
Comment (zll600) [reply t/quic/admin/basic.t:86]: OK, I will change it.

File: apisix/init.lua [hunk 8]
```diff
@@ -576,6 +576,13 @@ end
 
 
 function _M.http_access_phase()
+    if ngx.req.http_version() == 3 then
+        local upstream_host = ngx.var.host
+        if ngx.var.server_port then
+            upstream_host = upstream_host .. ":" .. ngx.var.server_port
+        end
+        ngx.var.upstream_host = upstream_host
+    end
```
Comment (zll600) [inline apisix/init.lua:585]: For http3, the `http_post` variable will be empty which is used to [set `$upstream_host`](https://github.com/apache/apisix/blob/d6a3db3ee2738436a634c3a6addd02c9cdfca035/apisix/cli/ngx_tpl.lua#L722). But because http/1.1 is used between apisix and upstream, `upstream_host` needs to be set manually here.

File: apisix/cli/ops.lua [hunk 9]
```diff
@@ -449,7 +460,7 @@ Please modify "admin_key" in conf/config.yaml .
                 end
 
                 listen_table_insert(node_listen, "http", ip, port,
-                        enable_http2, enable_ipv6)
+                        enable_http2, false, enable_ipv6)
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:463]: The `enable_quic` option is fixed?

File: ci/common.sh [hunk 10]
```diff
@@ -177,7 +186,7 @@ GRPC_SERVER_EXAMPLE_VER=20210819
 
 linux_get_dependencies () {
     apt update
-    apt install -y cpanminus build-essential libncurses5-dev libreadline-dev libssl-dev perl libpcre3 libpcre3-dev libldap2-dev
+    apt install -y cpanminus build-essential libncurses5-dev libreadline-dev libssl-dev perl libpcre3 libpcre3-dev libldap2-dev xz-utils
```
Comment (monkeyDluffy6017) [inline ci/common.sh:189]: Why do we need `xz-utils`?

Comment (zll600) [reply ci/common.sh:189]: I need it to install `static-url` from the `.tar.xz` file. ref: https://github.com/apache/apisix/pull/10989/files#diff-67362488684645b13d1314d9d45bdfd21fed3599970720bccd584a079971dff9R80

File: .github/workflows/quic.yml [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+        run: |
+          echo "version=${GITHUB_REF##*/}" >>$GITHUB_OUTPUT
+          echo "fullname=apache-apisix-${GITHUB_REF##*/}-src.tgz" >>$GITHUB_OUTPUT
+
+      - name: Extract test type
+        shell: bash
+        id: test_env
+        run: |
+          test_dir="${{ matrix.test_dir }}"
+          if [[ $test_dir =~ 't/quic/plugin' ]]; then
```
Comment (shreemaan-abhishek) [inline .github/workflows/quic.yml:75]: why do we need this if we are just testing basic.t?

File: apisix/cli/ops.lua [hunk 12]
```diff
@@ -407,7 +413,12 @@ Please modify "admin_key" in conf/config.yaml .
 
             if ip_port_to_check[addr] == nil then
                 table_insert(listen_table,
-                        {ip = ip, port = port, enable_http2 = enable_http2})
+                        {
```
Comment (shreemaan-abhishek) [inline apisix/cli/ops.lua:416]: can you explain the usage of this block briefly? And why this change is needed.

File: ci/common.sh [hunk 13]
```diff
@@ -75,6 +75,15 @@ install_curl () {
     curl -V
 }
 
+install_static_curl () {
```
Comment (shreemaan-abhishek) [inline ci/common.sh:78]: if this is not going to be used anywhere else, better move it in quic.yaml.

File: ci/common.sh (hunk 10 above)
Comment (shreemaan-abhishek) [reply ci/common.sh:189]: I hope you checked if there is any other alternative way to install static-url? Or install other tool? Or avoid installing it altogether.

File: apisix/cli/ops.lua (hunk 12 above)
Comment (zll600) [reply apisix/cli/ops.lua:416]: because I want to add a new field `enable_quic`. If I add `enable_quic` directly, the length of this line will too long to cause a lint error.

File: ci/common.sh (hunk 13 above)
Comment (zll600) [reply ci/common.sh:78]: OK, will do

File: ci/common.sh (hunk 10 above)
Comment (zll600) [reply ci/common.sh:189]: I will install xz-utils in quic.yml to avoid installing `xz-utils` altogether.

File: .github/workflows/quic.yml (hunk 11 above)
Comment (zll600) [reply .github/workflows/quic.yml:75]: I will add `t/quic/plugin` in another PR after merging this PR. So I didn't adjust code here which was copied from build.yml.

File: conf/config-default.yaml [hunk 14]
```diff
@@ -97,9 +97,11 @@ apisix:
     listen:                                       # APISIX listening port for HTTPS traffic.
       - port: 9443
         enable_http2: true
+        enable_quic: false
```
[the remark this answers is not in the document]
Comment (zll600) [reply conf/config-default.yaml:100]: `enable_quic` is already under `ssl`. I will add some comments for `enable_quic`.

File: apisix/init.lua (hunk 8 above)
Comment (zll600) [reply apisix/init.lua:583]: It will lost the `port` field by only using `ngx.var.http_host`.
Now Could we keep `ngx.var.host + ngx.var.server_port`?

File: t/quic/admin/basic.t [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+--- request
+GET /t
+--- response_body
+passed
+
+
+
+=== TEST 4: Successfully access route with QUIC
+--- exec
+curl -k -v -H ":authority: test.com" -H "content-length: 0" --http3-only --resolve "test.com:1994:127.0.0.1" https://test.com:1994/hello 2>&1 | cat
```
Comment (membphis) [inline t/quic/admin/basic.t:106]: this should be wrong: ":authority: test.com", pls fix it

File: .github/workflows/quic.yml (hunk 11 above)
Comment (shreemaan-abhishek) [reply .github/workflows/quic.yml:75]: ok

File: ci/common.sh [hunk 16]
```diff
@@ -177,7 +177,7 @@ GRPC_SERVER_EXAMPLE_VER=20210819
 
 linux_get_dependencies () {
     apt update
-    apt install -y cpanminus build-essential libncurses5-dev libreadline-dev libssl-dev perl libpcre3 libpcre3-dev libldap2-dev
+    apt install -y cpanminus build-essential libncurses5-dev libreadline-dev libssl-dev perl libpcre3 libpcre3-dev libldap2-dev xz-utils
```
Comment (shreemaan-abhishek) [inline ci/common.sh:180]: you are already installing xz-utils in quic.yml. Is this needed again?

Comment (zll600) [reply ci/common.sh:180]: Sorry, I will revert this change

**Review by membphis — COMMENTED** [review]
LGTM

need  a better title, eg: `feat: support to enable quic and add basic test case`

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by membphis — APPROVED** [review]
LGTM
